docs/architecture/ADR-004-reserves.md

```diff
@@ -0,0 +1,129 @@
+
+## Alternative Approaches
+
+- Setting aside a separate account (or even separate module account, for example by making `x/reserve`) would separate reserve balances from the lending pool.
+
+## Decision
+
+The `x/leverage` module already has a module account capable of storing base asset types, and it currently does so for assets in the lending pool.
+
+Adding reserve functionality should be as simple as using the existing module account to store reserve assets, and tracking (for each stored denomination) the amount which cannot be moved from the module acount (by means of a borrower's Borrow or a lender's Withdraw) because it is reserved.
+
+The aforementioned restrictions on borrows and withdrawals must then be implemented.
+
+Additionally, a governance parameter must be kept which specifies the portion of borrow interest that will be funnelled into reserves. This parameter will either be decided per token, or as a single parameter for all asset types.
+
+> TODO: Decide on the above.
+
+Regarding the actual increasing of reserves when interest accrues on borrow positions, the following approach can be used (which interacts with the lender interest stream defined in [ADR-001](./ADR-001-interest-stream.md).)
+
+For each token type:
+
+- The `x/leverage` has a non-negative balance of the token in the module account
+- There is a non-negative sum of all open borrow positions using that token type, which can be calculated by the `x/leverage` module.
+- The `x/leverage` module records a non-negative amount of the token which is declared to be reserved.
+- There is also an associated uToken type, of which the total amount in circulation can be measured by the `x/bank` module.
+
+Then:
+
+- For every open borrow of every token type, interest accrues during EndBlock using a dynamically calculated interest rate.
+- At the same moment, a portion of that increase is applied to the amount of the token that is declared to be reserved.
+- The token:uToken exchange rate, which would have previously calculated by `(total tokens borrowed + module account balance) / uTokens in circulation` for a given token type and associated uToken denomination, must now be computed as `(total tokens borrowed + module account balance - reserved amount) / uTokens in circulation`.
+
+## Detailed Design
+
+The `x/leverage` module account can be used without modification to store reserves.
+
+The `x/leverage` module keeper must store, for each token denom, store an `sdk.Int` representing the amount of of that token which is reserved (i.e. may not be borrowed or withdrawn from the module account.) The keeper will work as follows:
+
+```go
+// pseudocode
+reservePrefix | tokenDenom = sdk.Int
+```
+
+The logic for increasing the reserved amount of each token will occur in the EndBlock function, simultaneous with interest accrual:
+
+```go
+// pseudocode
+for each borrow {
```

Comment:
For each borrow position or each borrowed denom? The former would be too expensive and costly to do per block. If we have to do it per borrow position, then we need to do this in epochs.

Reply:
> By "shouldn't be a problem" I meant that doing it every X blocks instead of one wouldn't negatively impact the overall design.

Ahh perfect!

> One thing we need when calculating interest is the amount of time (seconds, years, etc) that have elapsed since the last time interest was accrued.

Yes, you can do this. You have access to the block height and time.